### PR TITLE
feat: upgrade core to OCaml 5.2

### LINKED ISSUE #1071: OCaml 5.2 support
The current version of melange fails with the following error when using OCaml 5.2:
```
#=== ERROR while compiling melange.3.0.0-51 ===================================#
# context              2.2.0~beta2~dev | linux/x86_64 | ocaml-variants.5.2.0+trunk | file:///home/opam/opam-repository
# path                 ~/.opam/5.2/.opam-switch/build/melange.3.0.0-51
# command              ~/.opam/5.2/bin/dune build -p melange -j 1 @install
# exit-code            1
# env-file             ~/.opam/log/melange-20-0841f9.env
# output-file          ~/.opam/log/melange-20-0841f9.out
### output ###
# (cd _build/default && /home/opam/.opam/5.2/bin/ocamlc.opt -w -9 -g -bin-annot -I vendor/melange-compiler-libs/lib/.melange_compiler_libs.objs/byte -I /home/opam/.opam/5.2/lib/menhirLib -I /home/opam/.opam/5.2/lib/ocaml/compiler-libs -I vendor/melange-compiler-libs/wrapper/.melange_wrapper.objs/byte -no-alias-deps -open Melange_compiler_libs__ -o vendor/melange-compiler-libs/lib/.melange_compiler_libs.objs/byte/melange_compiler_libs__Misc.cmi -c -intf vendor/melange-compiler-libs/lib/misc.mli)
# File "vendor/melange-compiler-libs/utils/misc.mli", line 469, characters 15-47:
# Error: Unbound type constructor "Melange_wrapper.Misc.Color.color"
# (cd _build/default && /home/opam/.opam/5.2/bin/ocamlc.opt -w -40 -w -69 -g -bin-annot -I jscomp/js_parser/.js_parser.objs/byte -no-alias-deps -open Js_parser -o jscomp/js_parser/.js_parser.objs/byte/js_parser__Flow_ast_mapper.cmo -c -impl jscomp/js_parser/flow_ast_mapper.ml)
# File "jscomp/js_parser/flow_ast_mapper.ml", line 1906, characters 7-12:
# 1906 |     [@@alert deprecated "Use either function_expression or class_method"]
#               ^^^^^
# Warning 53 [misplaced-attribute]: the "alert" attribute cannot appear in this context
# (cd _build/default && /home/opam/.opam/5.2/bin/ocamlopt.opt -w -40 -w -69 -g -O3 -unbox-closures -I jscomp/js_parser/.js_parser.objs/byte -I jscomp/js_parser/.js_parser.objs/native -intf-suffix .ml -no-alias-deps -open Js_parser -o jscomp/js_parser/.js_parser.objs/native/js_parser__Flow_ast_mapper.cmx -c -impl jscomp/js_parser/flow_ast_mapper.ml)
# File "jscomp/js_parser/flow_ast_mapper.ml", line 1906, characters 7-12:
# 1906 |     [@@alert deprecated "Use either function_expression or class_method"]
#               ^^^^^
# Warning 53 [misplaced-attribute]: the "alert" attribute cannot appear in this context
# (cd _build/.sandbox/18ba18a7722dac8844b705e4c060f6d2/default && /home/opam/.opam/5.2/bin/menhir --lalr --explain --dump --require-aliases --strict --table -lg 1 -la 1 --unused-token COMMENT --unused-token DOCSTRING --unused-token EOL --unused-token GREATERRBRACKET --fixed-exception vendor/melange-compiler-libs/lib/parser.mly --base vendor/melange-compiler-libs/lib/parser --infer-read-reply vendor/melange-compiler-libs/lib/parser__mock.mli.inferred)
# Grammar has 208 nonterminal symbols, among which 15 start symbols.
# Grammar has 123 terminal symbols.
# Grammar has 859 productions.
# Built an LR(0) automaton with 1875 states.
# The grammar is not SLR(1) -- 140 states have a conflict.
# The construction mode is lalr.
# Built an LR(1) automaton with 1875 states.
# 1069 shift/reduce conflicts were silently solved.
```

### PR DESCRIPTION
**Pull request by anmonteiro**
- adapts to the upcoming version of OCaml, 5.2
- this work is a companion to the PR in melange-compiler-libs: https://github.com/melange-re/melange-compiler-libs/pull/38
- while odoc and ocamlformat don't have a compatible release (see the OCaml 5.2 readiness issue https://github.com/ocaml/opam-repository/issues/25182), we'll stick to 5.1 for development and keep the compiler libs PR open

fixes #1071

### CODE REVIEW

File: vendor/dune [hunk 1]
```diff
@@ -22,4 +40,6 @@
    (package melange)
    (flags -w -9)
    (libraries menhirLib melange_wrapper)
-   (modules_without_implementation annot))))
+   (modules_without_implementation
+    annot
+    (:include ../gen/optional_5_2_modules.txt)))))
```
Comment (anmonteiro) [inline vendor/dune:45]: this is the reason why we have to bump the dune lang version to 3.13

Comment (anmonteiro) [reply vendor/dune:45]: https://github.com/ocaml/dune/pull/9578

File: .github/workflows/opam-build.yml [hunk 2]
```diff
@@ -26,6 +26,7 @@ jobs:
           # - windows-latest # for some reason windows build is not picking dune 3.5
         ocaml-compiler:
           - ocaml-base-compiler.5.1.1
+          # - 'ocaml-base-compiler.5.2.0~alpha1'
```
Comment (jchavarri) [inline .github/workflows/opam-build.yml:29]: Will this be uncommented before merging?

File: bin/melc.ml [hunk 3]
```diff
@@ -10,6 +10,7 @@
 (*                                                                     *)
 (***********************************************************************)
 
+open Melange_compiler_libs
```
Comment (jchavarri) [inline bin/melc.ml:13]: Why was this needed?

File: jscomp/core/js_implementation.cppo.ml [hunk 4]
```diff
@@ -90,19 +96,24 @@ let after_parsing_sig ppf outputprefix ast =
...
       let sg =
         let alerts = Builtin_attributes.alerts_of_sig ast in
-        Env.save_signature ~alerts tsg.Typedtree.sig_type modulename
+        Env.save_signature ~alerts tsg.Typedtree.sig_type
+#if OCAML_VERSION >= (5,2,0)
+          (Unit_info.make ~check_modname:false ~source_file:fname outputprefix
+           |> Unit_info.cmi)
+#else
+          modulename
           (Artifact_extension.append_extension outputprefix Cmi)
```
Comment (jchavarri) [inline jscomp/core/js_implementation.cppo.ml:105]: Why can't `Artifact_extension` be used in 5.2?

File: jscomp/js_parser/flow_ast_mapper.ml [hunk 5]
```diff
@@ -1903,7 +1903,7 @@ class ['loc] mapper =
         updated to override those individually. *)
     method function_expression_or_method loc (stmt : ('loc, 'loc) Ast.Function.t) =
       this#function_ loc stmt
-    [@@alert deprecated "Use either function_expression or class_method"]
+    (* [@@alert deprecated "Use either function_expression or class_method"] *)
```
Comment (jchavarri) [inline jscomp/js_parser/flow_ast_mapper.ml:1906]: Why?

File: playground/mel_playground.cppo.ml [hunk 6]
```diff
@@ -186,14 +186,23 @@ let compile =
     let types_signature = ref [] in
     warnings_collected := [];
     try
+      let lexbuf = Lexing.from_string str in
+      Ocaml_common.Location.input_lexbuf := Some lexbuf;
+      Location.input_lexbuf := Some lexbuf;
```
Comment (jchavarri) [inline playground/mel_playground.cppo.ml:191]: I saw these used in `Ast_io` too, is this something new that 5.2 requires? And does it affect somehow the older versions?

**Review by jchavarri — COMMENTED** [review]
I mostly have questions 🙃

File: .github/workflows/opam-build.yml (hunk 2 above)
Comment (anmonteiro) [reply .github/workflows/opam-build.yml:29]: No. The compiler libs submodule is on 5.1 according to the proposal I shared: have melange remain on 5.1 but ready to integrate the 5.2 submodule once OCamlformat and odoc have compatible releases.

File: jscomp/js_parser/flow_ast_mapper.ml (hunk 5 above)
Comment (anmonteiro) [reply jscomp/js_parser/flow_ast_mapper.ml:1906]: 5.2 errors on these misplaced attributes https://github.com/ocaml/ocaml/pull/12451

File: playground/mel_playground.cppo.ml (hunk 6 above)
Comment (anmonteiro) [reply playground/mel_playground.cppo.ml:191]: This is related to https://github.com/ocaml/ocaml/pull/12991

File: bin/melc.ml (hunk 3 above)
Comment (anmonteiro) [reply bin/melc.ml:13]: otherwise it doesn't know about coming from `Melange_compiler_libs.Unit_info`.

File: playground/mel_playground.cppo.ml (hunk 6 above)
Comment (anmonteiro) [reply playground/mel_playground.cppo.ml:191]: basically if we don't do this, tests won't include source code quotations.
